### ISSUE TITLE
Memory leak using cats.effect.std.Queue

### ISSUE THREAD

**Issue by paoloboni** [typelevel/fs2#2568]
Hi, I have some code that leaks memory and I'm not sure if it's a bug or just me making some wrong assumptions on the usage of `cats.effect.std.Queue` with fs2.

**Context**
The code below is a simplified version of some code I'm using to communicate with a websocket server. 
I'm using STTP in combination with Fs2 (see [here](https://sttp.softwaremill.com/en/latest/examples.html#open-a-websocket-using-fs2-streams)). The STTP API allows the interaction with the server by using a `fs2.Pipe`. Since in this case I'm only interested in the data flowing from the server to the client, what I'm trying to do is to define a `fs2.Stream` of events server -> client that can be easily consumed by some client code.

**Versions**
Fs2: 3.1.0
cats-effect: 3.2.2

Please assume the producer code cannot be modified, it allows processing elements of the source stream using a `fs2.Pipe`.
In that pipe I'm feeding a queue, that is later used to create the final stream.

```scala
import cats.effect.std.Queue
import cats.effect.{ExitCode, IO, IOApp}

import scala.util.Random
import fs2.{Pipe, Stream}

import scala.concurrent.duration.DurationInt

object SampleApp extends IOApp {
  val random = new Random()
  case class Something(a: String, b: Int, c: BigDecimal)

  val sampleData = Iterator.continually(
    Something(
      a = random.nextString(20),
      b = random.nextInt(),
      c = random.nextDouble()
    )
  )

  def producer(pipe: Pipe[IO, Something, Unit]): IO[Unit] = Stream
    .awakeEvery[IO]((50 + random.nextInt(50)).millis)
    .zipRight(Stream.fromIterator[IO](sampleData, 1))
    .through(pipe)
    .compile
    .drain

  def pipe(q: Queue[IO, Option[Something]]): Pipe[IO, Something, Unit] = { input =>
    input.evalMap(e => q.offer(Some(e)))
  }

  override def run(args: List[String]): IO[ExitCode] = (for {
    q <- Queue.unbounded[IO, Option[Something]]
    _ <- producer(pipe(q)).start
    _ <- Stream.fromQueueNoneTerminated(q).evalMap(e => IO.pure(println(e))).compile.drain
  } yield ()).redeem(
    { t =>
      t.printStackTrace()
      ExitCode(1)
    },
    _ => ExitCode.Success
  )
}
```

I tried running this process limiting the heap memory to 32Mb (`-Xmx32m`) and could see it failing with OOM error in about 30 min.

**Memory usage**
<img width="1041" alt="Screenshot 2021-08-22 at 11 57 08" src="https://user-images.githubusercontent.com/4746479/130352814-4f680704-b587-4d99-88c2-e31c0b51c117.png">

**Comment by Swoorup** [comment]
This originally came from investigating and could have cause this issue:
https://github.com/paoloboni/binance-scala-client/issues/133

**Comment by mpilquist** [comment]
I ran this program on Scala 2.13 with Java 11 and was unable to reproduce the leak. Heap usage did not grow at all. I also compared heap snapshots with YourKit and there was no growth.

The same program but modified to awake every nanosecond and also modified to discard the outputs of the queue (via `.evalMap(_ => IO.unit)` instead of printing them also experienced no heap growth.

One thing that jumps out to me is the use of an unbounded queue. If for some reason, the producer producers faster than the consumer consumes, you'll have unbounded growth. One way this could occur is if standard out was backed up -- e.g., running this program in a terminal and then scrolling back in history. I'd replace the unbounded queue with a bounded one to rule out this possibility.

If you can reproduce, grab a heap dump, perhaps by adding `-XX:+HeapDumpOnOutOfMemoryError`.

**Comment by paoloboni** [comment]
@mpilquist thanks for the quick reply. 
You are right, the code above works fine. What I actually tried locally was with this line [1]
```scala
_ <- Stream.fromQueueNoneTerminated(q).map(println).compile.drain
```
instead of [2]:
```scala
_ <- Stream.fromQueueNoneTerminated(q).evalMap(e => IO.pure(println(e))).compile.drain
```
Apologies for the confusion, it's something I thought to change while I was pasting the code in this issue since it didn't look very idiomatic, but now I'm realising it also makes a difference at runtime.

This is the memory dump when running the code above using [1]:
[java_pid33109.hprof.zip](https://github.com/typelevel/fs2/files/7027891/java_pid33109.hprof.zip)

**Comment by diesalbla** [comment]
Analysing that heap dump it with Eclipse inspector. 
 
> One instance of "cats.effect.IOFiber" loaded by "jdk.internal.loader.ClassLoaders$AppClassLoader @ 0x7fe00d5e8" occupies 27,230,184 (85.76%) bytes.

**Comment by diesalbla** [comment]
So, I have also been running the code test above, by copying it into a file in the `core` tests, and running it on top of current `main` version. It was running for a while, with reduced periods and even smaller heap (24 MiB), but I could not see it crash after a while.

**Comment by mpilquist** [comment]
Thanks, confirmed that this expression leaks:

```scala
Stream.fromIterator[IO](Iterator.continually(()), 1).map(identity).compile.drain
```

**Comment by mpilquist** [comment]
Adding this test to `MemoryLeakSpec` shows the issue:

```scala
leakTest("eval + flatMap + map (2)".only) {
  def unfold: Stream[IO, Unit] = Stream.eval(IO.unit).flatMap(_ => Stream.emits(Nil) ++ unfold)
     unfold.map(x => x)
}
```

**Comment by diesalbla** [comment]
An update, with today's discussions from Discord. We have submitted a PR, #2569, that seems to fix this memory leak by removing the pull-to-pull rewrite. Unfortunately it introduces a different memory leak at another test, so we keep working on it.

**Comment by Swoorup** [comment]
Just to complete, the following also leaks. No `map`, the for-comprehension likely uses `map` under the hood.

<details>
<summary>Details</summary>

```scala
import cats.effect.std.Queue
import cats.effect.{ExitCode, IO, IOApp}
import scala.util.Random
import fs2.{Pipe, Stream}
import scala.concurrent.duration.DurationInt

object SampleApp extends IOApp {

  val random = new Random()

  case class Something(a: String, b: Int, c: BigDecimal)

  val sampleData = Iterator.continually(
    Something(
      a = random.nextString(20),
      b = random.nextInt(),
      c = random.nextDouble()
    )
  )

  def producer(pipe: Pipe[IO, Something, Unit]): IO[Unit] = Stream
    .awakeEvery[IO]((5 + random.nextInt(5)).millis)
    .zipRight(Stream.fromIterator[IO](sampleData, 1))
    .through(pipe)
    .compile
    .drain

  def pipe(q: Queue[IO, Option[Something]]): Pipe[IO, Something, Unit] = { input =>
    input.evalMap(e => q.offer(Some(e)))
  }

  def fromProducer: Stream[IO, Something] = 
    Stream
      .eval(for {
        q <- Queue.bounded[IO, Option[Something]](10)
        _ <- producer(pipe(q)).void.start
      } yield Stream.fromQueueNoneTerminated(q))
      .flatten

  def consume: Stream[IO, Something] = 
    for {
      stream <- fromProducer
    } yield stream

  override def run(args: List[String]): IO[ExitCode] = (for {
    _ <- consume.evalMap(_ => IO.unit).compile.drain
  } yield ()).redeem(
    { t =>
      t.printStackTrace()
      ExitCode(1)
    },
    _ => ExitCode.Success
  )
}
```

</details>

**Comment by mpilquist** [comment]
Fixed in 3.1.1 which was released today
